hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/Reconciler.java

```diff
@@ -0,0 +1,110 @@
+package org.apache.mesos.hdfs.scheduler;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.mesos.hdfs.state.IPersistentStateStore;
+import org.apache.mesos.Protos.TaskState;
+import org.apache.mesos.Protos.TaskStatus;
+import org.apache.mesos.Protos;
+import org.apache.mesos.SchedulerDriver;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Observable;
+import java.util.Observer;
+import java.util.Set;
+
+/**
+ * HDFS Mesos Framework Reconciler class implementation.
+ */
+public class Reconciler implements Observer {
+  private final Log log = LogFactory.getLog(HdfsScheduler.class);
+  private IPersistentStateStore store;
+  private Set<String> pendingTasks;
+
+  public Reconciler(IPersistentStateStore pStore) {
+    store = pStore;
+    pendingTasks = new HashSet<String>();
+  }
+
+  public void reconcile(SchedulerDriver driver) {
+    pendingTasks = store.getAllTaskIds();
+    logPendingTasks();
```

Comment:
Thanks very much for the explanation. It makes sense, assuming this has been tested. Let's test a few scenarios here. Shutdown down the scheduler. Take down a NN by killing the executor process. Start the scheduler see if recovers. Same for DN and JN. If these basic tests pass, we should be good!